support quoted strings in HTTPHeaders[canonicalForm:]

tanner0101 opened the pull request:
Adds support for quoted strings in HTTPHeaders[canonicalForm:]

### Motivation:

Fixes https://github.com/apple/swift-nio/issues/1441

### Modifications:

Improves HTTP header value parsing.

### Result:

HTTP headers now respects quoted strings.

tanner0101 commented:
I've got a working implementation here but I'm not 100% sure this is the best approach. I'm interested for your thoughts.

Lukasa commented on Sources/NIOHTTP1/HTTPTypes.swift:
```diff
@@ -472,7 +472,114 @@ public struct HTTPHeaders: CustomStringConvertible, ExpressibleByDictionaryLiter
+
+    private mutating func pop() {
+        if self.current.startIndex != self.current.endIndex {
+            self.current = self.current[self.current.index(after: self.current.startIndex)...]
```
Just write `self.current = self.current.dropFirst()`.

Lukasa commented on Sources/NIOHTTP1/HTTPTypes.swift:
```diff
@@ -472,7 +472,114 @@ public struct HTTPHeaders: CustomStringConvertible, ExpressibleByDictionaryLiter
+        guard let nextQuote = self.current[startIndex...].firstIndex(of: "\"") else {
+            return nil
+        }
+        if self.current[self.current.index(before: nextQuote)] == "\\" {
```
This is not guaranteed to be safe: if the double-quote is literally at the front of the current string then this will perform an out-of-bounds access. Can you add a test-case that includes parsing the quoted string `""`? It should fail here, and we can then fix it.

Lukasa commented on Sources/NIOHTTP1/HTTPTypes.swift:
```diff
@@ -472,7 +472,114 @@ public struct HTTPHeaders: CustomStringConvertible, ExpressibleByDictionaryLiter
+        }
+        if self.current[self.current.index(before: nextQuote)] == "\\" {
+            // Skip escaped quotes.
+            return self.nextDoubleQuote(from: self.current.index(after: nextQuote))
```
It would be better not to recurse here: can we replace this with a loop-based solution instead?

Lukasa commented on Sources/NIOHTTP1/HTTPTypes.swift:
```diff
@@ -472,7 +472,114 @@ public struct HTTPHeaders: CustomStringConvertible, ExpressibleByDictionaryLiter
+        } else if let comma = self.nextComma() {
+            value = self.pop(to: comma)
+            self.pop()
+        } else if let whitespace = self.nextWhitespace() {
```
This...does not look right. What's the purpose of this clause?

Lukasa commented on Sources/NIOHTTP1/HTTPTypes.swift:
```diff
@@ -472,7 +472,114 @@ public struct HTTPHeaders: CustomStringConvertible, ExpressibleByDictionaryLiter
+private extension Substring {
+    /// Converts all `\"` to `"`.
+    func unescapingQuotes() -> Substring {
+        return self.split(separator: "\\").reduce("") { (result, part) -> SubSequence in
```
We should probably do this lazily to avoid taking an allocation hit here.

Lukasa replied:
We can also use `reduce(into:)` which should give us slightly nicer perf characteristics as well.

tanner0101 replied:
Any pointers on how this could be done lazily?

Lukasa replied:
`.lazy` would do the trick I think. ;)

tanner0101 commented:
@Lukasa thanks for the feedback. I've incorporated all of it. I'm not sure if I've used `lazy` correctly but it seems to work? I initially did `self.lazy.split` and that broke everything. But `self.split(...).lazy` works. This is my first time using that.

Lukasa commented on Sources/NIOHTTP1/HTTPTypes.swift:
```diff
@@ -472,7 +472,108 @@ public struct HTTPHeaders: CustomStringConvertible, ExpressibleByDictionaryLiter
+private extension Substring {
+    /// Converts all `\"` to `"`.
+    func unescapingDoubleQuotes() -> Substring {
+        return self.split(separator: "\\").lazy.reduce(into: "") { (result, part) in
```
`self.lazy.split.reduce` seems to work fine for me, and it's the correct spelling.

tanner0101 replied:
OK, cool. Thanks. I've taken this PR out of draft.

Lukasa replied:
Sorry, I think this meant moving the `lazy` to the left.

tanner0101 replied:
Oh sorry, I can't read. Lemme try that again.

tanner0101 replied:
Ah okay, I found the problem. The lazy subsequence's `CustomStringConvertible` conformance was different. I fixed this by using concatenation instead.

Lukasa commented:
@swift-nio-bot test this please

tanner0101 commented:
@Lukasa hold off on merging this actually. I'm realizing this implementation has a problem. Take this example:

```http
Foo: "a;b"; c="d;e", f
```

`HTTPHeaders[canonicalForm:]` would need to retain the quotes here in order for further parsing to be successful. I see a few options listed in order of preference: 

1: This PR is changed to allow `canonicalForm` to continue to respect quotes, but _without_ removing them from the result.
2: We abandon this PR and leave it up specific header field value parsers to figure out what works best for them. This means implementing their own version of canonical form parsing. 
3: We add a more advanced parser that is aware of directives / key-value pairs. 

(1) seems like the best option. (2) seems reasonable enough given `HTTPHeaders[canonicalForm:]` works for what NIO needs it for. (3) would require new API and seems possibly too opinionated to come from NIO. However, you could argue that this is quite difficult to get right and having NIO do that for you means less people will get it wrong.

Lukasa commented:
I'm strongly disinclined to do (3) in the abstract: I'd rather we provided an implementation of [Structured Headers](https://httpwg.org/http-extensions/draft-ietf-httpbis-header-structure.html) instead. Many pre-existing header field values fit into the structured headers format anyway, and it has the advantage of being a serious future-proofing move.

That makes me a bit inclined to do (2) and then implement Structured Headers instead.

tanner0101 commented:
@Lukasa that sounds good to me. Let's go with (2) then. Thanks for the link to Structured Headers, I'll research that and see if I can help with an implementation. 

Closing this PR in favor of https://github.com/apple/swift-nio/issues/1453
